packages/workspace/src/tasks-runner/hasher.ts

```diff
@@ -0,0 +1,87 @@
+import { ProjectGraph, ProjectGraphNode } from '../core/project-graph';
+import { NxJson } from '../core/shared-interfaces';
+import { Task } from './tasks-runner';
+import { createHash } from 'crypto';
+import { readFileSync, statSync } from 'fs';
+
+export class Hasher {
+  implicitDependenciesHash: string;
+  hashes: { [k: string]: string } = {};
+
+  constructor(
+    private readonly projectGraph: ProjectGraph,
+    private readonly nxJson: NxJson
+  ) {
+    this.implicitDependenciesHash = this.implicitDepsHash();
+  }
+
+  hash(task: Task): string {
+    const hash = createHash('sha256');
+    hash.update(task.target.project || '');
+    hash.update(task.target.target || '');
+    hash.update(task.target.configuration || '');
+    hash.update(JSON.stringify(task.overrides));
+    hash.update(this.implicitDependenciesHash);
+    this.traverseInDepthFirstOrder(task).forEach(p => {
+      hash.update(this.hashProjectNode(p));
+    });
+    return hash.digest('hex');
+  }
+
+  private traverseInDepthFirstOrder(task: Task): ProjectGraphNode[] {
+    const r = this.traverseNode(task.target.project, []);
+    return r.map(rr => this.projectGraph.nodes[rr]);
+  }
+
+  private traverseNode(project: string, acc: string[]): string[] {
+    if (acc.indexOf(project) > -1) return acc;
+    acc.push(project);
+    (this.projectGraph.dependencies[project] || [])
+      .map(t => t.target)
+      .forEach(r => {
+        this.traverseNode(r, acc);
+      });
+    return acc;
+  }
+
+  private hashProjectNode(p: ProjectGraphNode): string {
+    if (this.hashes[p.name]) {
+      return this.hashes[p.name];
+    } else {
+      const hash = createHash('sha256');
```

Comment:
Maybe we can hash the file stream instead of having to read the contents in? This would make it faster and less memory intensive. The downside is that the hasher will have to become async.

This lib uses workers if they are present so if things are slower we could steal some ideas from there: https://github.com/sindresorhus/hasha/blob/master/index.js

Reply:
I might just use hasha here. The hasher is used in async functions, so making hash async justifies the fact that we can use workers.